### PR TITLE
SLO-Doc Tracking For SLO-metric-data

### PR DESCRIPTION
**Pull request by bkez322**
For easy reading, please read [the actual proposal design-doc](
https://github.com/bkez322/dashdotdb/blob/multi-slo-doc-proposal/docs/design/slo_doc_tracking_for_slo_metric_data.md#slo-doc-tracking-for-slo-metric-data)

This PR adds a proposal doc based on internal Jira ticket [APPSRE-3570](https://issues.redhat.com/browse/APPSRE-3570), and is intended to solve [this problem](https://issues.redhat.com/browse/APPSRE-3570?focusedCommentId=18878164&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-18878164).

### CODE REVIEW

**Comment by bkez322** [comment]
Here is the actual proposal design-doc:

https://github.com/bkez322/dashdotdb/blob/multi-slo-doc-proposal/docs/design/multi_slo_doc_support.md

File: docs/design/multi_slo_doc_support.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  - ...
+```
+
+Visual-Qontract will have a new column `Doc Name` in the 'Service SLO' table.
+
+### Schema Changes
+
+Here are three proposals for how a 'slo_doc_name' identifier can be introduced to the schema:
+* [Option 1] (least effort): Update the 'ServiceSLO' table to include a 'SLODocName' column of type 'string'.
+* [Option 2] (author's recommendation): Update the 'ServiceSLO' table to include a 'SLODocName' column of type 'integer', which is a foreign key to a new table 'SLODocName'. Table 'SLODocName' contains 'id' (primary key, integer), and 'name' (string, unique)
```
Comment (patjlm) [inline docs/design/multi_slo_doc_support.md:66]: I am unsure of the value brought by externalizing the doc name in it own table here. If this table contains only an id and a name, and is only foreign-keyed from the ServiceSLO table, it would be simpler and easier to manage to do the option 1.

Comment (patjlm) [reply docs/design/multi_slo_doc_support.md:66]: Also, if we go for this option, the new table name should probably be more `SLODoc` rather than `SLODocName`

File: docs/design/multi_slo_doc_support.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+The service ['service-registry'](https://gitlab.cee.redhat.com/service/app-interface/-/tree/6762ce9ae3c34abedc21abf768437a5e7753652a/data/services/service-registry/slo-documents) would have [a similar problem](https://visual-app-interface.devshift.net/reports#/reports/service-registry/2021-08-01.yml)(although its visual report page fails to render for a seemingly-unrelated-reason) to 'ocm'.
+
+## Proposal
+
+### Overview
+
+The 'name' ([example](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L7)) of these SLO docs will be introduced as an identifier for the SLO metric data stored in dashdotdb.
```
Comment (patjlm) [inline docs/design/multi_slo_doc_support.md:45]: Just a warning here on the API. Introducing a new field here would make the API non-backward compatible if the field is mandatory in queries (esp. GET I guess).
Either we accept that because we control all the API clients, or we need to make this backward compatible in some ways: allow queries without SLODoc and return as today, fail if several record match (from several SLODoc)

File: docs/design/multi_slo_doc_support.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+### Schema Changes
+
+Here are three proposals for how a 'slo_doc_name' identifier can be introduced to the schema:
+* [Option 1] (least effort): Update the 'ServiceSLO' table to include a 'SLODocName' column of type 'string'.
+* [Option 2] (author's recommendation): Update the 'ServiceSLO' table to include a 'SLODocName' column of type 'integer', which is a foreign key to a new table 'SLODocName'. Table 'SLODocName' contains 'id' (primary key, integer), and 'name' (string, unique)
+    * Better data normalization that Option 1
+    * Less disruptive to schema than Option 3
+* [Option 3] (most effort): Same as Option 2, and in addition: update all tables that currently have a relationship with table 'ServiceSLO' to instead have a relationship with 'SLODocName'
+
+[This PR](https://github.com/app-sre/dashdotdb/pull/50) uses the "Option 2" approach.
```
Comment (rrati) [inline docs/design/multi_slo_doc_support.md:71]: Rather than link to a PR for implementation of an idea, please spell out the details of each proposal (if needed for clarity) here.  This should be independent of an implementation, and is usually done before any implementation is done.

File: docs/design/multi_slo_doc_support.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+This digram visualizes the proposed schema change, introducing the 'SLODocName' table. (bottom left):
+![](images/multi-slo-doc-schema-change.png)
+
+### Other Changes To DashDotDB 
+
+* dash.db
+  * this needs a "slo_doc_name" property: https://github.com/app-sre/dashdotdb/blob/f88048cd9156afa93a4a0448ecd088c551c8408d/dashdotdb/models/dashdotdb.py
```
Comment (rrati) [inline docs/design/multi_slo_doc_support.md:79]: This is pointing to the whole file.  I assume you mean to point at a specific line (or set of lines)?  Probably the ServiceSLO model?

File: docs/design/multi_slo_doc_support.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  * update this to include "slo_doc_name": https://github.com/app-sre/dashdotdb/blob/f88048cd9156afa93a4a0448ecd088c551c8408d/examples/serviceslometrics.json
+  * update this image: https://github.com/app-sre/dashdotdb/blob/f88048cd9156afa93a4a0448ecd088c551c8408d/docs/dashdotdb.png (and the diagram file for it: https://github.com/app-sre/dashdotdb/blob/f88048cd9156afa93a4a0448ecd088c551c8408d/docs/dashdotdb.dia)
+  * Follow the [DB upgrade guide](https://github.com/app-sre/dashdotdb#db-upgrade) to upgrade the schema in Postgres
+
+### Changes To Other Systems
+
+The following are changes to related systems that are expected to need to be made.
+
+* app-interface
+  * add "slo_doc_name": https://gitlab.cee.redhat.com/service/app-interface/-/blob/15d1ce504a9233a04d6119e8d57e58a3bcc31cb0/graphql-schemas/schema.yml#L1916-1922
```
Comment (rrati) [inline docs/design/multi_slo_doc_support.md:89]: Why is this needed?  The SLODocument_v1 already has a name field here.  The SLO document names are in the SLODocumentSLO_v1 schema.

File: docs/design/multi_slo_doc_support.md (hunk 2 above)
Comment (rrati) [reply docs/design/multi_slo_doc_support.md:45]: The name linked here is the name of the service.  Do you want the name of the slo doc?  That would be in the list of slos in the app-interface schema

File: docs/design/multi_slo_doc_support.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+## Proposal
+
+### Overview
+
+The 'name' ([example](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L7)) of these SLO docs will be introduced as an identifier for the SLO metric data stored in dashdotdb.
+
+SLO data will be stored in report-YML files in the following format:
+```
+service_slo:
+  slo_doc_name1:
```
Comment (rrati) [inline docs/design/multi_slo_doc_support.md:50]: Is there a desire to split out the name of the slo like this instead of just having service_slo be a list similar to what it is in app-interace?

File: docs/design/multi_slo_doc_support.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  * update this to include 'slo_doc_name': https://gitlab.cee.redhat.com/service/app-interface/-/blob/15d1ce504a9233a04d6119e8d57e58a3bcc31cb0/schemas/app-sre/slo-document-1.yml
+  * All of the SLO documents here need to have the new 'slo_doc_name' proeprty set: https://gitlab.cee.redhat.com/service/app-interface/-/tree/5381889ea1aecd1d5afd51fae246c6a5909438ef/data/services
+  * all of these same sort of changes in public-github app-interface
+
+* visual-qontract
+  * This needs to accept MULTIPLE 'slo_documents' as a parameter https://github.com/app-sre/visual-qontract/blob/bb5e993a428780ccd6a1d4208b36740e46d96f79/src/pages/elements/Report.js#L547
+  * This must pass MULTIPLE 'slo_documents': https://github.com/app-sre/visual-qontract/blob/bb5e993a428780ccd6a1d4208b36740e46d96f79/src/pages/elements/Report.js#L704
+  * This needs to be reethought to grab multiple slo documents, rather than just 1: https://github.com/app-sre/visual-qontract/blob/bb5e993a428780ccd6a1d4208b36740e46d96f79/src/pages/elements/Report.js#L633-L643
```
Comment (rrati) [inline docs/design/multi_slo_doc_support.md:97]: Do you have ideas on how to do this?  Being this vague isn't going to helpful to understand what was changed because of this proposal

**Comment by rrati** [comment]
Please be more explicit with regards to the proposed changers.  There's a lot in this that is hard to follow because there isn't proper context in the document about what and how things will be changed.

See https://github.com/app-sre/dashdotdb/blob/master/docs/design/explict_tokens.md as an example for being more explicit.

File: docs/design/multi_slo_doc_support.md (hunk 1 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md:66]: 'Option 1' is certainly easier. The externalized table was a bit of a future-proofing exercise, but quite happy to plan for 'Option 1' if we don't think that's necessary.

Otherwise, 'SLODoc' would probably be a more appropriate table name, sure.

File: docs/design/multi_slo_doc_support.md (hunk 2 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md:45]: > That would be in the list of slos in the app-interface schema

@rrati 

Can you link an example of what you mean?

Are you referring to [this](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L16)?

Comment (rrati) [reply docs/design/multi_slo_doc_support.md:45]: https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L13

File: docs/design/multi_slo_doc_support.md (hunk 4 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md:79]: Yes, sorry, that link was supposed to be

https://github.com/app-sre/dashdotdb/blob/f88048cd9156afa93a4a0448ecd088c551c8408d/dashdotdb/models/dashdotdb.py#L168

File: docs/design/multi_slo_doc_support.md (hunk 5 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md:89]: Thank you for pointing this out. It is not needed.

I had intended to delete this point.

File: docs/design/multi_slo_doc_support.md (hunk 6 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md:50]: @rrati 

Can you give an example of the output you're proposing?

Like this?

```
service_slo:
  - cluster: app-sre-prod-01
    slo_doc_name: some-doc-name-1
    namespace: app-interface-production
    slo_name: Reconciliation time of app-interface integrations
    slo_value: 27.0
    slo_target: 95.0
  - cluster: app-sre-prod-01
    slo_doc_name: some-doc-name-2
    namespace: app-interface-production
    slo_name: Reconciliation time of app-interface integrations
    slo_value: 27.0
    slo_target: 95.0
  - ...
```

If so that should be fine.

Comment (rrati) [reply docs/design/multi_slo_doc_support.md:50]: Yep, something like that is what I'm proposing.

File: docs/design/multi_slo_doc_support.md (hunk 7 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md:97]: I do need to spend more time on thinking through the code changes for visual-qontract.

My intended scope with this design-doc was primarily just around dashdotDB changes, with some high-level references to related systems.

If desired, I'm happy to take more time to update this document to more holistically cover the specific code changes to all related systems.

Comment (rrati) [reply docs/design/multi_slo_doc_support.md:97]: If it's not related to DDDB, I wouldn't include it tbh.  This design doc should just be about changes to DDDB and independent of other services.  Those other services will need their down design docs for how to consume the changes in DDDB.

Of course, the changes here need to be done in a way that can be consumed by the users of the system, but how they will do it really isn't a concern of the DDDB proposal.

File: docs/design/multi_slo_doc_support.md (hunk 2 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md:45]: @rrati 

Perhaps I am severely misunderstanding something.

You're saying [this](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L7) isn't the "name" of an [SLO doc](https://gitlab.cee.redhat.com/service/app-interface/-/blob/15d1ce504a9233a04d6119e8d57e58a3bcc31cb0/graphql-schemas/schema.yml#L1916)?

Isn't what you just linked the name of the `SLO` (defined [here](https://gitlab.cee.redhat.com/service/app-interface/-/blob/15d1ce504a9233a04d6119e8d57e58a3bcc31cb0/graphql-schemas/schema.yml#L1904)) and **not** the name of the `SLO doc` (defined [here](https://gitlab.cee.redhat.com/service/app-interface/-/blob/15d1ce504a9233a04d6119e8d57e58a3bcc31cb0/graphql-schemas/schema.yml#L1920))?

My understanding is that [this](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L13)(your link) and [this](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/clusters-service.yml#L13) use identical SLO names for the same service (ocm)... which is problematic for the way [reports are produced for this service](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/reports/ocm/2021-08-01.yml#L128-143).

File: docs/design/multi_slo_doc_support.md (hunk 6 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md:50]: Cool. I think I prefer that too.

File: docs/design/multi_slo_doc_support.md (hunk 7 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md:97]: Fair enough. I will delete the section with references to high-level changes to other systems from this particular design doc.

File: docs/design/multi_slo_doc_support.md (hunk 2 above)
Comment (rrati) [reply docs/design/multi_slo_doc_support.md:45]: I think you're right.  I linked to the name of the SLO vs the name of the slo document.  My bad.

**Comment by bkez322** [comment]
> Please be more explicit with regards to the proposed changers. There's a lot in this that is hard to follow because there isn't proper context in the document about what and how things will be changed.
> 
> See https://github.com/app-sre/dashdotdb/blob/master/docs/design/explict_tokens.md as an example for being more explicit.

My apologies.

This initial draft of this design doc had a lot of focus on the big picture problem experienced by end-users, and how dash.db, app-interface, visual-qontract, and qontract-reconcile all tie into that problem together.

I suppose TL;DR version of this design-doc should be rewritten to simply be:

"I propose that 'SLO metric data' being stored in dashdotDB be changed to look like [this](https://github.com/app-sre/dashdotdb/blob/16584a71b9e751583a762a0433852e2bbd98dfa8/examples/serviceslometrics.json#L3), and this is how to accomplish that"

I will plan to improve the focus on that specific point, and remove many details I wrote on app-interfact, visual-qontract, and qontract-reconcile.

File: docs/design/slo_doc_tracking_for_slo_metric_data.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+---
+
+[`GET /api/v1/serviceslometrics`](https://github.com/app-sre/dashdotdb/blob/4663b71f34b2c540f7163985b7eccae8e7a9b0f0/dashdotdb/schemas/swagger.yaml#L208-L240)
+
+The JSON response body now includes a 'slo_doc_name' property.
+
+Example: *(same as above)*
+
+In addition, 'slo_doc_name' is added as an optional query parameter. The rationale for this being optional is to avoid breaking existing clients that are functioning fine without burdening them with needing to know a 'slo_doc_name'. A known risk with making this optional, however, is that clients potentially may make queries and get results from a 'slo_doc_name' that they don't intend to be querying from.
```
Comment (rrati) [inline docs/design/slo_doc_tracking_for_slo_metric_data.md:80]: I think it is worse to keep backwards compatibility but serve potentially incorrect data than to make a breaking change and ensure served data is correct.  We control everything that talks to DDDB atm so we can catch everything that has issues.  The intention is to correct the data problem, so imo we should accept that fixing the problem is a breaking change.

We had to do something similar when we introduced tokens.  We added non-optional fields because that was the only way to do things the right way.

File: docs/design/slo_doc_tracking_for_slo_metric_data.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+---
+
+## Implementation - Database Schema Changes
+
+What follow are two proposals for how a 'slo_doc_name' identifier can be introduced to the schema. The document-author reccomends 'Option 2' for the sake of schema consistency. Changes are marked in green.
```
Comment (rrati) [inline docs/design/slo_doc_tracking_for_slo_metric_data.md:86]: typos:
follow -> following
reccomends -> recommends

**Comment by rrati** [comment]
@bkez322 Well done! Much improved.  Some minor nitpicks, but otherwise looks good.  I favor the 2nd option as well.  In addition to what you mentioned as reasons, it also make slo_doc similar to sli_type.

I would prefer @rporres take a look as well to ensure nothing is missed on the DB schemas.  Once we agree on an implementation you should remove the option we don't choose before we accept and merge.

File: docs/design/slo_doc_tracking_for_slo_metric_data.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    "value": 66,
+    "target": 99
+}
+```
+
+---
+
+[`GET /api/v1/serviceslometrics/metrics`](https://github.com/app-sre/dashdotdb/blob/4663b71f34b2c540f7163985b7eccae8e7a9b0f0/dashdotdb/schemas/swagger.yaml#L193-L207)
+
+The objects within the JSON response body array will now include a 'slo_doc_name' property.
```
Comment (bkez322) [inline docs/design/slo_doc_tracking_for_slo_metric_data.md:68]:
```suggestion
The objects within the JSON response body array will now include a 'slo_doc.name' property.
```

File: docs/design/slo_doc_tracking_for_slo_metric_data.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+Example: *(same as above)*
+
+---
+
+[`GET /api/v1/serviceslometrics`](https://github.com/app-sre/dashdotdb/blob/4663b71f34b2c540f7163985b7eccae8e7a9b0f0/dashdotdb/schemas/swagger.yaml#L208-L240)
+
+The JSON response body now includes a 'slo_doc_name' property.
```
Comment (bkez322) [inline docs/design/slo_doc_tracking_for_slo_metric_data.md:76]:
```suggestion
The JSON response body now includes a 'slo_doc.name' property.
```

File: docs/design/slo_doc_tracking_for_slo_metric_data.md (hunk 9 above)
Comment (bkez322) [reply docs/design/slo_doc_tracking_for_slo_metric_data.md:86]:
```suggestion
What follows are two proposals for how a 'slo_doc_name' identifier can be introduced to the schema. The document-author recommends 'Option 2' for the sake of schema consistency. Changes are marked in green.
```

File: docs/design/multi_slo_doc_support.md (hunk 1 above)
Comment (rporres) [reply docs/design/multi_slo_doc_support.md]: I do agree that we could do it with Option 1, but definitely is more in line with the overall schema design to put it in its own table, especially since we can have multiple SLO definitions relating to one SLO document, so I agree with Option 2 in this case. DB normalization rules are always somewhat obscure to me, but I guess that should be to make it satisfy the 2NF.

File: docs/design/slo_doc_tracking_for_slo_metric_data.md (hunk 8 above)
Comment (rporres) [reply docs/design/slo_doc_tracking_for_slo_metric_data.md]: Sorry for coming late to the party, but I agree with @rrati. We control the clients for dashdotdb, so I would not make it optional and introduce a breaking change.

**Comment by rporres** [comment]
1000 apologies again for reviewing this so late. The document is very clear and easy to follow. I've commented above in the main two doubts: the db normalization (or not) and the breaking change API.

File: docs/design/multi_slo_doc_support.md (hunk 1 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md]: Unless further discussion arises, I'll plan on utilizing 'Option 2' for this.

File: docs/design/multi_slo_doc_support.md (hunk 2 above)
Comment (bkez322) [reply docs/design/multi_slo_doc_support.md]: Discussion has shifted to: https://github.com/app-sre/dashdotdb/pull/51#discussion_r701909921

**Review by rporres — APPROVED** [review]

File: docs/design/slo_doc_tracking_for_slo_metric_data.md (hunk 8 above)
Comment (bkez322) [reply docs/design/slo_doc_tracking_for_slo_metric_data.md]: Doc has been updated to reflect this as a required property.
